grails-app/services/au/org/ala/ecodata/UserService.groovy

```diff
@@ -179,55 +166,44 @@ class UserService {
     }
 
 
-    /**
-     * Get user from JWT.
-     * @param authorizationHeader
-     * @return
-     */
-    au.org.ala.web.UserDetails getUserFromJWT(String authorizationHeader = null) {
-        if((config == null) || (alaOidcClient == null))
-            return
-        try {
-            GrailsWebRequest grailsWebRequest = GrailsWebRequest.lookup()
-            HttpServletRequest request = grailsWebRequest.getCurrentRequest()
-            HttpServletResponse response = grailsWebRequest.getCurrentResponse()
-            if (!authorizationHeader)
-                authorizationHeader = request?.getHeader(AUTHORIZATION_HEADER_FIELD)
-            if (authorizationHeader?.startsWith("Bearer")) {
-                final WebContext context = FindBest.webContextFactory(null, config, JEEContextFactory.INSTANCE).newContext(request, response)
-                def optCredentials = alaOidcClient.getCredentials(context, config.sessionStore)
-                if (optCredentials.isPresent()) {
-                    Credentials credentials = optCredentials.get()
-                    def optUserProfile = alaOidcClient.getUserProfile(credentials, context, config.sessionStore)
-                    if (optUserProfile.isPresent()) {
-                        def userProfile = optUserProfile.get()
-                        String userId = userProfile?.userId ?: userProfile?.getAttribute(grailsApplication.config.getProperty('userProfile.userIdAttribute'))
-                        if (userId) {
-                            return authService.getUserForUserId(userId)
-                        }
-                    }
-                }
-            }
-        } catch (Throwable e) {
-            log.error("Failed to get user details from JWT", e)
+    private String checkForDelegatedUserId(HttpServletRequest request) {
+        // When BioCollect or MERIT calls ecodata, they use a M2M access token which contains custom scopes to
+        // enable access to the ecodata API.
+        // We can trust the requests containing bearer tokens with this scope and extract the userId from a header.
+        String scope = grailsApplication.config.getProperty('app.readScope')
+        if (!scope) {
+            log.error("No read scope specified in config.")
             return null
         }
+
+        if (request.isUserInRole(scope)) {
```

Comment:
does isUserInRole check scope attribute or just role attribute?

Reply:
Yes, for M2M tokens it applies the role check to the scopes rather than the user profile roles.  The alternatives I could find to identify M2M was to check the principal (which is set to the clientId) or a check of the profile class (which introduces a dependency on the plugin implementation).  
I thought since we otherwise were using the scopes it made sense to go with that approach?

Reply:
(The disadvantage of checking the client id is we need to manage a list of authorized clients)

Reply:
Thanks, looks good to me. I was just checking.